microshift_running_apps/microshift-greenboot.adoc

```diff
@@ -0,0 +1,89 @@
+:_content-type: ASSEMBLY
+[id="microshift-greenboot"]
+= The greenboot health check
+include::_attributes/attributes-microshift.adoc[]
+:context: microshift-greenboot
+
+toc::[]
+
+You might be challenged troubleshooting edge-device problems after software or an operating system upgrade fails. Edge devices are often in locations where direct serviceability is either limited or non-existent. In these cases, you can use greenboot to troubleshoot software problems for these devices. Greenboot is the Generic Health Check Framework for the `systemd` service on RPM-OSTree-based systems.
+
+Using greenboot can reduce your risk of being locked out of an edge device during upgrades and also prevent a significant interruption of service if an upgrade fails. When a failure is detected, the system boots into the last known working configuration using the `rpm-ostree` rollback capability.
+
+include::modules/microshift-install-greenboot-rpm.adoc[leveloffset=+1]
+include::modules/microshift-greenboot-healthchecks-list.adoc[leveloffset=+1]
+
+The `40_microshift_running_check.sh` health check script is also included in the `microshift-greenboot` RPM. The script validates that the required {product-title} services are running.
+//Q is the microshift script in addition to the fedora greenboot rpms, or are we giving users just the microshift-greenboot rpm?
+
+Exiting the script with a non-zero status will have the boot declared as failed. Greenboot redirects all the script output to the system log, accessible via the `journalctl -u greenboot-healthcheck.service` command.
+
+.Validation statuses and outcome for {product-title}.
+
+[cols="3", options="header"]
+|===
+|Validation
+|Pass
+|Fail
+
+|Check the script runs with 'root' permissions
+|Next
+|exit 0
+
+|Check microshift.service is enabled
+|Next
+|exit 0
+
+|Wait for microshift.service to be active (!failed)
+|Next
+|exit 1
+
+|Wait for Kubernetes API health endpoints to be OK
+|Next
+|exit 1
+
+|Wait for any Pod to start
+|Next
+|exit 1
+
+|For each core namespace, wait for images to be pulled
+|Next
+|exit 1
+
+|For each core namespace, wait for Pods to be ready
+|Next
+|exit 1
+
+|For each core namespace, check Pods not restarting
+|exit 0
+|exit 1
+|===
+
+[NOTE]
+====
+If the system is not booted using an OS-Tree file system, the health check procedures still run, but no rollback is possible in case of an upgrade failure.
+====
+
+The wait period in each validation is a minimum of five minutes. This minimum wait period is incrementally increased by the base wait period after each boot in the verification loop. You can override the base time wait period by setting the MICROSHIFT_WAIT_TIMEOUT_SEC environment variable in the `/etc/greenboot/greenboot.conf` configuration file.
+//Q: does this mean each validation check represented by a row of the table above takes 5 minutes? this section is confusing
```

Comment:
Each validation waits at most MICROSHIFT_WAIT_TIMEOUT_SEC for it to be successful. Once successful, the script proceeds with the next validation immediatelly. The timeout determines how long to wait for a validation to succeed before declaring it a failure.